Smarter field caps with subscribable listener

astefan opened the pull request:
This change is two fold:
- refactors `EsqlSession` to use (hopefully) a more readable and manageable series of async calls that start with pre-analysis of enrich policies, regular indices, lookup indices and end with verification of the analyzed logical plan of the query. It's changing the nested calls using ActionListeners to a "flat" series of steps using `SubscribableListener`
- in case of `VerificationException` being thrown in the verification step (following the analysis of the logical plan) and given a REST request filter was used, it removes the filter and retries the analysis. Assuming the first call to `IndexResolver` using the filter (and implicitly to `_field_caps` call where the filter is being used, as well) fails, it's retries the `_field_caps` call without a filter.

There is a leftover task (to be addressed in a separate PR) that should add tests in CCS scenarios: https://github.com/elastic/elasticsearch/issues/118054

Addresses https://github.com/elastic/elasticsearch/issues/115053

astefan committed:
WIP

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_fieldcaps_calls

astefan committed:
wip

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
whatevs

astefan commented:
@elasticmachine run elasticsearch-ci/part-2

astefan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into smarter_field_caps_subscribableListener

astefan committed:
Small fix, big impact

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Merge branch 'smarter_field_caps_subscribableListener' of https://github.com/astefan/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Fix some things and rewrite the listeners a bit

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan commented:
@dnhatn @original-brownbear asking for a double-checking of how I tried to use the SubscribableListener.
There is still something not right with my approach, given the tests that fail... one of them is about a VerificationException that is not thrown from the two-three places in this PR where I catch them (while calling `analyzeAction.apply(indexResolution, enrichResolution)`). I wasn't aware that specific exception is thrown from other places other than the Verifier. This proves to a degree that the way the listeners are chained and how they deal with this type of exception is not 100% correct.

astefan committed:
Link the original Listener to the Subscribable one, as well

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Small fix

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
wip

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into smarter_field_caps_subscribableListener

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -544,4 +630,52 @@ public PhysicalPlan optimizedPhysicalPlan(LogicalPlan optimizedPlan) {
         return plan;
     }
+
+    class IndexResolutionWrappingListener implements ActionListener<EnrichResolution> {
+
+        private final ActionListener<Tuple<EnrichResolution, IndexResolution>> delegate;
+        private final IndexResolution indexResolution;
+
+        IndexResolutionWrappingListener(
+            ActionListener<Tuple<EnrichResolution, IndexResolution>> delegate,
+            IndexResolution indexResolution
+        ) {
+            this.delegate = delegate;
+            this.indexResolution = indexResolution;
+        }
+
+        @Override
+        public void onResponse(EnrichResolution enrichResolution) {
+            delegate.onResponse(new Tuple<>(enrichResolution, indexResolution));
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            delegate.onFailure(e);
+        }
+    }
+
+    class EnrichResolutionWrappingListener implements ActionListener<IndexResolution> {
+
```
These two classes are very similar - use a base class with Generics to pass the parameter in.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -330,22 +337,96 @@ private <T> void preAnalyze(
+                // whatever tuple we have here (from CCS-special handling or from the original pre-analysis), pass it on to the next step
+                l.onResponse(tuple);
+            })
+            .<Tuple<EnrichResolution, IndexResolution>>andThen((l, tuple) -> {
```
👍
This style is more readable.
Furthermore I would extract the lambda in each stage in a separate method - it improves readability and avoids casts (inferred based on the method signature).

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -544,4 +630,52 @@ public PhysicalPlan optimizedPhysicalPlan(LogicalPlan optimizedPlan) {
+        private final IndexResolution indexResolution;
+
+        IndexResolutionWrappingListener(
+            ActionListener<Tuple<EnrichResolution, IndexResolution>> delegate,
```
Using a record instead of Tuple<En...> makes sense here and allows for future evolution without noise all over the file:
`record ListenerResult(EnrichResolution, IndexResolution,...)` since we're likely have more `Resolution` in the future

costin reviewed (commented):
👍

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Fix after update from main

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Merge branch 'smarter_field_caps_subscribableListener' of https://github.com/astefan/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Put back code that existed before update from main

astefan committed:
Small fix
Align ListenerResult params with AnalyzerContext params

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
spotless

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -302,81 +303,113 @@ private <T> void preAnalyze(
+            })
+            .<ListenerResult>andThen((l, listenerResult) -> {
+                // first attempt (maybe the only one) at analyzing the plan
+                analyzeAndMaybeRetry(analyzeAction, requestFilter, listenerResult, l, logicalPlanListener);
```
Apart from the refactoring of EsqlSession to use a SubscribableListener, this is the meat of the change: retrying the analysis step in case a VerificationException (which, with few exceptions, is being thrown from the Verifier that runs as part of the `Analyzer.analyze` call) happens on the first attempt.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -409,38 +442,117 @@ private void preAnalyzeIndices(
+                } else {
+                    // interested only in a VerificationException, but this time we are taking out the index filter
+                    // to try and make the index resolution work without any index filtering. In the next step... to be continued
+                    l.onResponse(listenerResult);
```
I _think_ we should return here. Otherwise `logicalPlanListener` will be called with a `null` plan. 
(And I guess that call can be moved within the `try` and the `return`s dropped.)

astefan replied:
I am not sure, tbh. Seems logical what you suggest and I made the change, but CI will tell us if something breaks. I almost 100% guided myself with the correctness of the changes by observing CI.

bpintea reviewed (commented):
First try to understand the change.

astefan committed:
Add IT tests

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -156,7 +159,8 @@ public void execute(EsqlQueryRequest request, EsqlExecutionInfo executionInfo, P
                 }
-            }
+            },
+            request.filter()
```
The functionality in this PR is about the request filter. I haven't considered and I don't think it's correct either to take the filtering from the query itself (like `where some_field > 1`) and put it inside the `field_caps` call. The validity of the `where` command depends on what fields can be found in the index pattern from `from` command. If the index pattern is filtered by the `where` command I think there is a sort of "loop" between field names resolution and query validity.

astefan replied:
Yep, that's perfectly fair. And an improvement nonetheless; I have my reservations regarding how obvious the "filters" in a query are considered eligible for also filtering the columns.

dnhatn commented:
@astefan Would you mind updating the PR's description? I think I'm missing the context of the issue we're trying to solve. Thanks!

astefan committed:
Don't run the tests in async as well; it's too unpredictable

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Update docs/changelog/116755.yaml

elasticsearchmachine commented:
Hi @astefan, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

astefan commented:
> @astefan Would you mind updating the PR's description? I think I'm missing the context of the issue we're trying to solve. Thanks!

Sorry, @dnhatn, you're right. I've added tests, updated the description of the PR and marked as ready for review.

astefan commented:
@elasticmachine run elasticsearch-ci/part-1

astefan committed:
Merge branch 'main' into smarter_field_caps_subscribableListener

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -409,38 +442,117 @@ private void preAnalyzeIndices(
+        LogicalPlan plan = null;
+        var filterPresentMessage = requestFilter == null ? "without" : "with";
+        var attemptMessage = requestFilter == null ? "the only" : "first";
+        LOGGER.debug("Analyzing the plan ({} attempt, {} filter)", attemptMessage, filterPresentMessage);
```
Imo, it's an oversimplification. I'd rather not do it.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -302,81 +303,113 @@ private <T> void preAnalyze(
+            })
+            .<LogicalPlan>andThen((l, listenerResult) -> {
+                assert requestFilter != null : "The second analysis shouldn't take place when there is no index filter in the request";
+                LOGGER.debug("Analyzing the plan (second attempt, without filter)");
+                LogicalPlan plan;
+                try {
+                    plan = analyzeAction.apply(listenerResult.indices, listenerResult.lookupIndices, listenerResult.enrichResolution);
+                } catch (Exception e) {
+                    l.onFailure(e);
+                    return;
+                }
+                LOGGER.debug("Analyzed plan (second attempt, without filter):\n{}", plan);
+                l.onResponse(plan);
```
I am not sure, tbh. Seems logical what you suggest and I made the change, but CI will tell us if something breaks. I almost 100% guided myself with the correctness of the changes by observing CI.

astefan committed:
Address reviews

astefan committed:
Merge branch 'main' of https://github.com/elastic/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Merge branch 'smarter_field_caps_subscribableListener' of https://github.com/astefan/elasticsearch into smarter_field_caps_subscribableListener

astefan committed:
Minor change

astefan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into smarter_field_caps_subscribableListener

costin reviewed (approved):
LGTM

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 116755`
